clippy_lints/src/zombie_processes.rs

```diff
@@ -0,0 +1,253 @@
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::visitors::for_each_local_use_after_expr;
+use clippy_utils::{fn_def_id, get_enclosing_block, match_any_def_paths, match_def_path, paths};
+use rustc_ast::Mutability;
+use rustc_hir::def::DefKind;
+use rustc_hir::intravisit::{walk_expr, Visitor};
+use rustc_hir::{Expr, ExprKind, HirId, Local, Node, PatKind, Stmt, StmtKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::{sym, Span};
+use std::ops::ControlFlow;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Looks for code that spawns a process but never calls `wait()` on the child.
+    ///
+    /// ### Why is this bad?
+    /// As explained in the [standard library documentation](https://doc.rust-lang.org/stable/std/process/struct.Child.html#warning),
+    /// calling `wait()` is necessary on Unix platforms to properly release all OS resources associated with the process.
+    /// Not doing so will effectively leak process IDs and/or other limited global resources,
+    /// which can eventually lead to resource exhaustion, so it's recommended to call `wait()` in long-running applications.
+    /// Such processes are called "zombie processes".
+    ///
+    /// ### Example
+    /// ```rust
+    /// use std::process::Command;
+    ///
+    /// let _child = Command::new("ls").spawn().expect("failed to execute child");
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::process::Command;
+    ///
+    /// let mut child = Command::new("ls").spawn().expect("failed to execute child");
+    /// child.wait().expect("failed to wait on child");
+    /// ```
+    #[clippy::version = "1.74.0"]
+    pub ZOMBIE_PROCESSES,
+    suspicious,
+    "not waiting on a spawned child process"
+}
+declare_lint_pass!(ZombieProcesses => [ZOMBIE_PROCESSES]);
+
+fn emit_lint(cx: &LateContext<'_>, span: Span) {
+    span_lint_and_then(
+        cx,
+        ZOMBIE_PROCESSES,
+        span,
+        "spawned process is never `wait()`-ed on and leaves behind a zombie process",
```

Comment:
Should mention this only happens on unix (or might leave behind), a line break would also be appreciated

Reply:
can you elaborate on what you mean by line break?

Reply:
Making it span 2 lines, it's a bit long for one imo. Not necessary though

Reply:
You mean 2 lines for the *message*? Does that actually work? Never seen that, but I'm also not really sure if that's what you mean (probably not). I suppose "might leave behind a zombie process" could be its own note. Then the message is shorter: just "spawned process is never `wait()`-ed on"

Reply:
Now that I think about it, I don't think I have either. Yea that's what I meant. A separate note is good too :+1: